go/service/gregor_test.go

```diff
@@ -178,3 +188,266 @@ func TestShowTrackerPopupMessage(t *testing.T) {
 		t.Fatalf("Expected the tracker window for UID %s to be dismissed. current value: %s", trackee.User.GetUID().String(), identifyUI.dismissedUsername)
 	}
 }
+
+func newMsgID() gregor1.MsgID {
+	ret := make([]byte, 16)
+	rand.Read(ret)
+	return ret
+}
+
+type mockGregord struct {
+	sm  gregor.StateMachine
+	fc  clockwork.FakeClock
+	log logger.Logger
+}
+
+func (m mockGregord) Sync(_ context.Context, arg gregor1.SyncArg) (gregor1.SyncResult, error) {
+	return grpc.Sync(m.sm, rpc.SimpleLogOutput{}, arg)
+}
+func (m mockGregord) ConsumeMessage(_ context.Context, msg gregor1.Message) error {
+	_, err := m.sm.ConsumeMessage(msg)
+	return err
+}
+func (m mockGregord) ConsumePublishMessage(_ context.Context, _ gregor1.Message) error {
+	return errors.New("unimplemented")
+}
+func (m mockGregord) Ping(_ context.Context) (string, error) {
+	return "pong", nil
+}
+
+func (m mockGregord) newIbm(uid gregor1.UID, msgid gregor1.MsgID) gregor1.Message {
+	m.fc.Advance(time.Minute)
+	return gregor1.Message{
+		Ibm_: &gregor1.InBandMessage{
+			StateUpdate_: &gregor1.StateUpdateMessage{
+				Md_: gregor1.Metadata{
+					Uid_:   uid,
+					MsgID_: msgid,
+					Ctime_: gregor1.ToTime(m.fc.Now()),
+				},
+				Creation_: &gregor1.Item{
+					Category_: "unknown!",
+					Body_:     gregor1.Body([]byte("HIHIHI")),
+				},
+			},
+		},
+	}
+}
+
+func newGregordMock(logger logger.Logger) mockGregord {
+	var of gregor1.ObjFactory
+	fc := clockwork.NewFakeClock()
+
+	sm := storage.NewMemEngine(of, fc)
+
+	return mockGregord{sm: sm, fc: fc, log: logger}
+}
+
+func setupSyncTests(t *testing.T, tc libkb.TestContext) (*gregorHandler, mockGregord, gregor1.UID) {
+	var err error
+	user, err := kbtest.CreateAndSignupFakeUser("gregr", tc.G)
+	if err != nil {
+		t.Fatal(err)
+	}
+	uid := gregor1.UID(user.User.GetUID().ToBytes())
+
+	var h *gregorHandler
+	if h, err = newGregorHandler(tc.G); err != nil {
+		t.Fatal(err)
+	}
+
+	server := newGregordMock(tc.G.Log)
+
+	return h, server, uid
+}
+
+func checkMessages(t *testing.T, source string, msgs []gregor.InBandMessage,
+	refMsgs []gregor.InBandMessage) {
```

Comment:
are we guaranteed no duplicates in these lists?  Because testing set length equality and set inclusion for set equality will only work in that case...